PackedScene: Avoid saving connections twice

cixil opened the pull request:
This PR is a redo of #100965

Fixes #48064 fixes #86532 fixes #85372 (variations of the same issue)

### The Problem
When packing a PackedScene from a script that contained instances of other scenes, signal connections within those instances were also getting saved in some circumstances (they should all be ignored because they are already saved in the scene that was instanced). This causes "duplicate signal" errors whenever the scenes were instantiated as the editor would try to connect the signals more than once. As some of the issues point out, you can open a `tscn` file and plainly see that connections are saved where they shouldn't be. The errors can quickly overwhelm the rest of the output and obfuscate real problems.

### What this PR does
This PR fixes the issues by giving connections "owners" when they are instantiated and checking the owners when packing them to ensure they remain only in the scenes that own them. The owner of a node is the scene it belongs to, but the root node of a scene instance will belong to the scene containing it. The new "connection owner" I have added is different because a connection from the root node of a scene instance (that belongs to the scene) should still be owned by the root node of the instance, not the root node of the containing scene.

Adding this new scheme made most of the previous connection finding code irrelevant, and I believe this method makes the whole connection packing and instantiation more straightforward.

cixil commented on scene/resources/packed_scene.cpp:
```diff
@@ -1160,51 +1171,12 @@ Error SceneState::_parse_connections(Node *p_owner, Node *p_node, HashMap<String
+			Node *conn_owner = p_node->get_connection_owner(target, E.name, base_callable);
+			bool owned_by_main_scene = (conn_owner == p_owner || conn_owner == nullptr); // nullptr means the connection was just added and wasn't connected during node instantiation.
 
-							if (from_node >= 0 && to_node >= 0) {
-								//this one has state for this node, save
-								if (state->is_connection(from_node, c.signal.get_name(), to_node, base_callable.get_method())) {
-									exists2 = true;
-									break;
-								}
-							}
-						}
-
-						nl = nullptr;
-					} else {
-						if (nl->is_instance()) {
-							Ref<SceneState> state = nl->get_scene_instance_state();
-							if (state.is_valid()) {
-								int from_node = state->find_node_by_path(nl->get_path_to(p_node));
-								int to_node = state->find_node_by_path(nl->get_path_to(target));
-
-								if (from_node >= 0 && to_node >= 0) {
-									//this one has state for this node, save
-									if (state->is_connection(from_node, c.signal.get_name(), to_node, base_callable.get_method())) {
-										exists2 = true;
-										break;
-									}
-								}
-							}
-						}
-						nl = nl->get_owner();
-					}
-				}
-
-				if (exists2) {
-					continue;
-				}
```
I could not see how the check for `exists2` covered any cases not covered by the check for `exists` and was not able to trigger line 1206, so this seems superfluous.

cixil replied:
but if it would be more secure to leave in the check for exists2 it would't change this PR.

KoBeWi commented:
Same as previous PR: https://github.com/godotengine/godot/pull/100965#issuecomment-2858308534

cixil commented:
It seems #86532 is actually a separate issue, I removed it from the PR description.
This PR solves an issue with `pack` saving connections it shouldn't and that issue is related only to instantiation.

cixil commented:
No, it is also to do with pack. I will look into it further.

cixil commented:
@KoBeWi I fixed it

KoBeWi commented on scene/main/node.h:
```diff
@@ -597,6 +605,25 @@ class Node : public Object {
+		data.connection_owners[owner] = sc;
+	}
+
+	Node *get_connection_owner(Node *to_node, const StringName &signal_name, Callable c) const {
```
```suggestion
	Node *get_connection_owner(Node *p_to_node, const StringName &p_signal_name, const Callable &p_callable) const {
```
Also why is the definition here, not in cpp?

cixil replied:
No reason. I'll move it.

cixil commented:
I think `CONNECT_INHERITED` could be removed then, but an additional function might need to be added for removing connection owners for "make local"

cixil commented:
I  went ahead and fixed it up for inherited connections too! I kept it as a second commit in case the original one is preferable.

cixil commented:
I fixed a couple small issues I noticed when using a build with this PR on my own project and made the connection ownership also work for inherited scenes so I could clean up the existing code more. 

It's ready for review when someone is available.

cixil commented:
Actually noticed another case where this doesn't work (when packing a node that is not in the tree). I'll need to work on this again when I have time.
```gdscript
func pack() -> void:
	var root = Node.new()
	# Some children have connections to eachother
	for child in get_children():
		child.reparent(root)
		child.owner = root
	
    #fixes the problem
	#add_child(root)
	#root.owner = owner
	
	var scene := PackedScene.new()
	scene.pack(root)
	ResourceSaver.save(scene,"res://test/saved_scene.tscn" )  # children's connections won't be saved
```

AThousandShips commented on scene/resources/packed_scene.cpp:
```diff
@@ -1197,93 +1229,17 @@ Error SceneState::_parse_connections(Node *p_owner, Node *p_node, HashMap<String
-							}
-						}
+			bool owned_by_node_in_tree = p_node == conn_owner || p_node->find_child(conn_owner->get_name(), true);
+			bool owned_by_main_scene = (conn_owner == p_owner || conn_owner == nullptr); // nullptr means the connection was just added and wasn't connected during node instantiation.
```
```suggestion
			bool owned_by_main_scene = conn_owner == p_owner || conn_owner == nullptr; // nullptr means the connection was just added and wasn't connected during node instantiation.
```

cixil committed:
avoid saving connections twice and clean up connection packing

KoBeWi commented on scene/main/node.cpp:
```diff
@@ -2586,6 +2586,27 @@ void Node::_propagate_reverse_notification(int p_notification) {
+void Node::add_connection_owner(Node *p_owner, Node *p_to_node, const StringName &p_signal_name, const Callable &p_callable, bool is_inherited) {
+	uint32_t hash = HashMapHasherDefault::hash(p_to_node) ^ p_signal_name.hash() ^ p_callable.get_method().hash();
+	if (is_inherited) {
+		// this connection exists from instantiation but is inherited, so return some pointer that
```
```suggestion
		// This connection exists from instantiation but is inherited, so return some pointer that
```
Applies to all comments.

KoBeWi reviewed (approved):
So as I said in https://github.com/godotengine/godot/pull/110969#issuecomment-3349424497, this fixes all of the issues now, but needs core review (reminder that I'm not in core team).

cixil commented:
I may have found another edgecase where this is still duplicating signals with editable instances. Going to try and look into/resolve today and then fix the feedback.

cixil commented on scene/main/node.cpp:
```diff
@@ -2586,6 +2586,27 @@ void Node::_propagate_reverse_notification(int p_notification) {
+	if (is_inherited) {
+		// this connection exists from instantiation but is inherited, so return some pointer that
+		// isn't the actual owner and isn't nullptr.
+		data.connection_owners[hash] = reinterpret_cast<Node *>(1);
```
This also has to work for nodes that aren't in the scenetree, so I'm not sure how to get the scenetree root in that case.

KoBeWi replied:
`SceneTree::get_singleton()->get_root()`

cixil replied:
thanks, I tried replacing the dummy pointer with that and the line in `packed_scene.cpp` 
```conn_owner = (Node*)SceneTree::get_singleton()->get_root() ```
 segfaults on the unit tests. I'm guessing the scenetree singleton doesn't exist in the test environment and there's some way to "enable" it for these tests, but that doesn't feel as clean.

cixil replied:
I could add a dummy pointer in the node class to point to for this purpose?

cixil replied:
By constant do you mean it should be a constant variable in the node class?

cixil replied:
if this is unsafe I could just maintain inherited connections in a separate list like connection_owners and check that instead.

cixil commented:
@KoBeWi other than fixing comments according to your feedback and adding tests, I only changed `p_node->find_child` to `p_owner->find_child` here in `packed_scene.cpp` to fix the issue I mentioned a couple hours ago.
-     bool owned_by_node_in_tree = p_node == conn_owner || (conn_owner != nullptr && p_node->find_child(conn_owner->get_name(), true));
+     bool owned_by_node_in_tree = p_node == conn_owner || (conn_owner != nullptr && p_owner->find_child(conn_owner->get_name(), true)); // Need to check this so that connections of reparented nodes not in the scenetree will be saved.

cixil commented:
Since I am a new contributor and this PR involves some refactoring, it seems unlikely that it will be considered anytime soon due to the [new contribution policy](https://godotengine.org/article/contribution-policy-2026/) (which is great btw). 

I still think this refactor is a good approach to the problem and would make the connection handling cleaner and more uniform. If a more experienced maintainer wants to take this idea and make a new PR or edit this one, by all means!

I may still update for 4.7 at some point anyway for my own project that is heavily affected by the bug.

Zireael07 commented:
@cixil I don't think the new contribution policy applies retroactively? Your PR was opened almost a year ago

KoBeWi commented:
Also it's a bugfix, so probably fine.

cixil commented:
This PR is ready to be reviewed again (no changes, just rebased to 4.8). I updated the description to make it more clear and will continue using a build of this PR when I use Godot.

KoBeWi commented:
#42161 was linked in the previous PR, but isn't in this one. Isn't it still being fixed?

cixil commented:
I am fairly confident that this would fix it, but I don't completely understand the problem in #42161 and there isn't an example project to test on.

I tried the example that Dragnaros92 posted in the last comment and wasn't able to reproduce exactly like they described when I opened in Godot 4.7 stable, however there were some duplicate errors printed just when looking at the level scene in the editor that went away using a build of this PR.

I think we could probably mark it resolved with this PR and add a comment that it wasn't actually tested?

siliconspecter commented:
Checked this out.  Seems to work perfectly - GDScript-generated PackedScenes are no longer duplicating the signals of child nodes which are themselves PackedScene instances.  I've also tested saving on a scene with over 2000 fully local signal connections and it didn't seem to do anything unexpected or produce any noticeable performance impact.
